homeassistant/components/template/template_entity.py

```diff
@@ -430,3 +445,220 @@ async def async_added_to_hass(self) -> None:
     async def async_update(self) -> None:
         """Call for forced update."""
         self._async_update()
+
+
+class TemplateRestoreEntity(RestoreEntity, TemplateEntity):
+    """Template Entity that restores data."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Template Restore Entity init."""
+        super().__init__(*args, **kwargs)
+        self._restore: bool = False
+        self._save_state: bool = True
+        self._additional_data: list[str] = []
```

Comment:
Can we just generate this on demand, instead of preparing it in advance?

Reply:
> On a general note: Restoring state is wanted for trigger-based template entities, but why is it wanted for state-based template entities as opposed to rendering a default value or setting the template entity unavailable?

Also having it for state based template entities allows one to keep existing state if the state of the other entity becomes unknown or unavailable.

For example:
```
- trigger:
      - platform: time_pattern
        # This will update every 10 minutes
        minutes: "/10"
    sensor:
      - name: hour
        unique_id: hour
        state: "{{ now().strftime('%M')|int(0) }}"
- sensor:
      - name: restore
        unique_id: restore
        state: >
          {{ (states('sensor.hour') not in ((None, 'None', 'unknown', 'unavailable')))|iif(states('sensor.hour'), states('sensor.restore')) }}
        attributes:
          last_update: "{{ (states('sensor.hour') not in ((None, 'None', 'unknown', 'unavailable')))|iif(states.sensor.hour.last_changed.strftime('%Y-%m-%d %H:%M'), state_attr('sensor.restore','last_update')) }}"
        restore: True
```

This is a simple example I used to test and technically could be made into a trigger template entity.
Upon reload (or restart) the entity sensor.hour will be unknown until minutes is divisible by 10. However, sensor.restore will be set to the last known state upon reload or restart. 
By using restore, state-based template entities can also be used to track values across restarts/reloads. 
For example, using it to calculate total value for entities that only provide delta values or by calculating delta value when a sensor only provided total value.
A restorable state-based template entity can also be used in this manner then to always provide the latest known value even if the underlying sensor becomes unavailable or unknown.